rust/src/download_compile.rs

```diff
@@ -0,0 +1,139 @@
+use crate::bs_safe::{BridgeStanError, Result};
+use flate2::read::GzDecoder;
+use log::info;
+use path_absolutize::Absolutize;
+use std::{
+    env::temp_dir,
+    fs,
+    path::{Path, PathBuf},
+};
+use tar::Archive;
+
+pub(crate) const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// Download and unzip the BridgeStan source distribution for this version
+/// to ~/.bridgestan/bridgestan-version
+pub fn get_bridgestan_src() -> Result<PathBuf> {
+    let homedir = dirs::home_dir().unwrap_or(temp_dir());
+
+    let bs_path_download_temp = homedir.join(".bridgestan_tmp_dir");
+    let bs_path_download = homedir.join(".bridgestan");
+
+    let bs_path_download_temp_join_version =
+        bs_path_download_temp.join(format!("bridgestan-{VERSION}"));
+    let bs_path_download_join_version = bs_path_download.join(format!("bridgestan-{VERSION}"));
+
+    if !bs_path_download_join_version.exists() {
+        info!("Downloading BridgeStan");
+
+        fs::remove_dir_all(&bs_path_download_temp).unwrap_or_default();
+        fs::create_dir(&bs_path_download_temp).unwrap_or_default();
+        fs::create_dir(&bs_path_download).unwrap_or_default();
+
+        let url = "https://github.com/roualdes/bridgestan/releases/download/".to_owned()
+            + format!("v{VERSION}/bridgestan-{VERSION}.tar.gz").as_str();
+
+        let response = ureq::get(url.as_str())
+            .call()
+            .map_err(|e| BridgeStanError::DownloadFailed(e.to_string()))?;
+        let len = response
+            .header("Content-Length")
+            .and_then(|s| s.parse::<usize>().ok())
+            .unwrap_or(50_000_000);
+
+        let mut bytes: Vec<u8> = Vec::with_capacity(len);
+        response
+            .into_reader()
+            .read_to_end(&mut bytes)
+            .map_err(|e| BridgeStanError::DownloadFailed(e.to_string()))?;
+
+        let tar = GzDecoder::new(bytes.as_slice());
+        let mut archive = Archive::new(tar);
+        archive
+            .unpack(&bs_path_download_temp)
+            .map_err(|e| BridgeStanError::DownloadFailed(e.to_string()))?;
+
+        fs::rename(
+            bs_path_download_temp_join_version,
+            &bs_path_download_join_version,
+        )
+        .map_err(|e| BridgeStanError::DownloadFailed(e.to_string()))?;
+
+        fs::remove_dir(bs_path_download_temp).unwrap_or_default();
+
+        info!("Finished downloading BridgeStan");
+    }
+
+    Ok(bs_path_download_join_version)
+}
+
+/// Compile a Stan Model given a stan_file and the path to BridgeStan
+/// if None, then calls get_bridgestan_src() to download BridgeStan
+pub fn compile_model<P>(
+    stan_file: P,
+    stanc_args: Vec<&str>,
+    make_args: Vec<&str>,
+    bs_path: Option<P>,
+) -> Result<PathBuf>
+where
+    P: AsRef<Path>,
+{
+    let bs_path = match bs_path {
+        Some(path) => path.as_ref().to_owned(),
+        None => get_bridgestan_src()?,
+    };
+
+    let stan_file = stan_file
+        .as_ref()
+        .absolutize()
```

Comment:
Why not https://doc.rust-lang.org/std/fs/fn.canonicalize.html
Couldn't we drop the path_absolutize dep then?

Reply:
this has some terrible "bug" on windows https://github.com/rust-lang/rust/issues/42869 It was hitting the CI/CD a few commits ago if you wish to check. I think std::path::absolute https://doc.rust-lang.org/std/path/fn.absolute.html would work here but is nightly.

Reply:
Oh no...
Maybe we could add a comment and mention that we can switch to https://doc.rust-lang.org/std/path/fn.absolute.html when that is stable?

Reply:
okay, added the comments